### ISSUE TITLE
Symfony 6.0 compatibility 

### ISSUE THREAD

**Issue by damienfa** [ecphp/cas-bundle#58]
## Description 

The Guards have been removed in Symfony 6.0. 
The `CasGuardAuthenticator` should be migrated to something like "CasCustomAuthenticator" using the new[ "custom authenticator"](https://symfony.com/doc/current/security/custom_authenticator.html).

## Expected Result

* Be compatible with Symfony 6.0 (and 5.4) 
* The new Authenticator should not be "final", it will be convenient to be able to "extends" in the code if needed.

**Comment by drupol** [comment]
Hello Damien,

Upgrading the packages for Symfony >= 5.4 has been discussed yesterday and it will come very soon.

Regarding the `final` keyword, I'm afraid that this won't change. If you're willing to "extend" a class, I would suggest you to use composition en inheritance.

We also did a presentation here : https://github.com/ecphp/session--composition-and-inheritance/

Feel free to let me know if you have any other questions.

**Comment by damienfa** [comment]
Ok, great about the update. Can't wait to use this library back. 👌🏼

About the final, I can perfectly use a composition (and I will). 

But, you make me wonder… I feel like my use-case is particularly adapted to inheritance and not composition : once the CAS has validated the user, I would like that my `InheritedCasAuthenticator` validates that the user has a valid registration on the product. For that, I just "add" protected/private methods (= to check the product registration) + overwrite one of the existing methods to call them and calling a `parent::` inside it. 
The semantic and mechanical aspect of the inheritance are respected. Isn't it? My subclass is still a CasAutenticator, but upgraded with a new feature. My inheritage is not done to "cherry pick" methods in order to reuse code, it's done reuse the whole class and its logic upon an upgraded CasAuthenticator.
In my case, why should you force to "composition pattern" the parent class with a `final` ?

**Comment by drupol** [comment]
Thanks :)

However, I'm using Composition over inheritance since year and I never found a use-case where inheritance was a better pattern.

If you want to customize the `validate` method of the `CasAuthenticator`, you have to create your own `Authenticator` and decorate the original one.

Some code worth ten thousands lines of explanation:

```php

// This is the original class
final class CasGuardAuthenticator extends AbstractGuardAuthenticator {
  // I omitted the body of it for brevity.
}

// My own custom decorator.
final class MyCustomCasGuardAuthenticator extends AbstractGuardAuthenticator {
  private AuthenticatorInterface $authenticator; // This is the decorated service.

  public function __construct(AuthenticatorInterface $authenticator) {
    $this->authenticator = $authenticator;
  }

    public function getCredentials(Request $request): ?ResponseInterface {
        // Call the original method here:
        $response = $this->authenticator->getCredentials($request);

       // Then do extra control here:
       return $this->checkIfUserHasValidRegistration($response);
    }

    private function checkIfUserHasValidRegistration(ResponseInterface $response): bool
    {
        return true; // Implements your custom business logic here.
    }

    // Implements all the rest of the methods and call the decorated authenticator
    // when you do not need to customize anything.
}
```

And don't forget to add this in `services.yaml` so the container will do it's job properly,
So that when you call the `cas.guardauthenticator`, your custom service will be served.

```yaml
services:
    FQDN\Of\My\Custom\GuardAuthenticator:
        decorates: 'cas.guardauthenticator'
        arguments: ['@.inner']
```

**Comment by drupol** [comment]
Yes, that's the price to pay and the only downside of using the decorator pattern. I actually could provide an abstract class which does most of the job, that's an idea I keep in my head for sure for the next iteration release.

**Comment by drupol** [comment]
@damienfa 

I created a branch that should fix the issue. I'm going to wait for some feedback of yours before merging anything.

The PR is sketchy, so it is possible that it might be broken, don't be afraid to report the least thing you see.

Tests are currently broken, I will fix them and the documentation in the meantime.

**Comment by drupol** [comment]
@damienfa Do you have any feedback to share so far?

**Comment by drupol** [comment]
Dear @Lubna93 ,

ecphp/cas-bundle is fully compatible with Symfony 6 and up now.

Can I close the issue ?

**Comment by Lubna93** [comment]
@drupol Thanks a lot for your great efforts! Your hard work to help others is so much  appreciated!
Yes, you can close it with many thanks!

**Comment by drupol** [comment]
You're very welcome :)
